ddtrace/contrib/celery/utils.py

```diff
@@ -39,40 +39,52 @@ def tags_from_context(context):
     return tags
 
 
-def attach_span(task, task_id, span):
+def attach_span(task, task_id, span, is_publish=False):
     """Helper to propagate a `Span` for the given `Task` instance. This
     function uses a `WeakValueDictionary` that stores a Datadog Span using
-    the `task_id` as a key. This is useful when information must be
+    the `(task_id, is_publish)` as a key. This is useful when information must be
     propagated from one Celery signal to another.
+
+    DEV: We use (task_id, is_publish) for the key to ensure that publishing a
+         task from within another task does not cause any conflicts.
+
+         This mostly happens when either a task fails and a retry policy is in place,
```

Comment:
Is there a test case for this you can add? If not, perhaps an explanation of why would be helpful.

Reply:
Done, I added a note that we cannot test this behavior well.